Migrate tuple(handle) -> handle

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105611
* #105606
* #105008
* #105007
* #105006
* __->__ #104488

We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping


cc @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

voznesenskym committed:
[WIP] Migrate tuple(handle) -> handle

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




[ghstack-poisoned]

wconstab commented on test/distributed/fsdp/test_fsdp_pure_fp16.py:
```diff
@@ -146,11 +146,11 @@ def _test_fp16_dtypes(
-            self.assertEqual(param.dtype, torch.float16)
-            if param.grad is not None:
-                self.assertEqual(param.grad.dtype, torch.float16)
+            # Check parameter/gradient dtypes
```
why is this indented?

wconstab commented on test/distributed/fsdp/test_fsdp_unshard_params.py:
```diff
@@ -288,8 +288,8 @@ def _test_unshard_params_respects_reshard(
         )
-        outer_flat_param = model._handles[0].flat_param
-        inner_flat_param = model.module[0]._handles[0].flat_param
+        outer_flat_param = model._handle[0].flat_param
```
should `_handle` still have a `[0]` after it?

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Migrate tuple(handle) -> handle"




cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -190,6 +189,8 @@ def _check_flat_params_on_expected_device(state: _FSDPState, module: nn.Module):
     """
     cpu_device = torch.device("cpu")
     for handle in traversal_utils._get_fsdp_handles(module):
+        if not handle:
```
I think we should get rid of this change since `_get_fsdp_handles()` filters the `None` handles.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -343,14 +346,14 @@ def _unshard(
     any_ran_pre_unshard = False
     with state._device_handle.stream(pre_unshard_stream):
-        for handle in handles:
+        if handle:
```
nit: We do not need this `if handle:` since we return early on `if not handle:`.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -360,64 +363,54 @@ def _unshard(
             event.synchronize()
     with state._device_handle.stream(unshard_stream):
-        for handle in handles:
+        if handle:
```
nit: Same here -- I do not think this is needed

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -360,64 +363,54 @@ def _unshard(
 ):
     """
-    Reshards the handles in ``handles``. ``free_unsharded_flat_params`` should
+    Reshards the handle. ``free_unsharded_flat_params`` should
     have the same length as ``handles``, and each element should give whether
     the corresponding handle should free its padded unsharded flat parameter.
```
```suggestion
    Reshards the handle. ``free_unsharded_flat_param`` indicates whether to
    free the handle's padded unsharded flat parameter.
```

awgu commented on torch/distributed/fsdp/_exec_order_utils.py:
```diff
@@ -135,15 +133,15 @@ def get_handles_to_forward_prefetch(
+        return target_handles_key
 
-    def record_post_forward(self, handles: List[FlatParamHandle]) -> None:
+    def record_post_forward(self, handle: FlatParamHandle) -> None:
```
```suggestion
    def record_post_forward(self, handle: Optional[FlatParamHandle]) -> None:
```
since we may actually pass `state._handle` directly, which may be `None`

awgu commented on torch/distributed/fsdp/_exec_order_utils.py:
```diff
@@ -153,19 +151,17 @@ def record_post_forward(self, handles: List[FlatParamHandle]) -> None:
-    def record_pre_forward(
-        self, handles: List[FlatParamHandle], is_training: bool
-    ) -> None:
+    def record_pre_forward(self, handle: FlatParamHandle, is_training: bool) -> None:
```
```suggestion
    def record_pre_forward(self, handle: Optional[FlatParamHandle], is_training: bool) -> None:
```
since we may actually pass `state._handle` directly, which may be `None`

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -512,35 +504,35 @@ def _post_forward(
-    free_unsharded_flat_params = [
-        not state._is_root
+    free_unsharded_flat_param = (
+        handle
```
nit: do not need the `handle` in this condition since we return early on `if not handle:`

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -616,9 +609,10 @@ def _root_pre_forward(
             handles_keys = []
             for fsdp_state in state._all_fsdp_states:
                 # TODO: Forward prefetch assumes singleton handles key. For the
-                # composable path, `_handles` may have more than one handle,
+                # composable path, `_handle` may have more than one handle,
                 # whereas for the wrapper path, it has at most one handle.
```
Let us delete this `TODO`.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -616,9 +609,10 @@ def _root_pre_forward(
             handles_keys = []
```
Let us rename this to `handles = []` (and change downstream).

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -1071,36 +1054,38 @@ def _finalize_params(
+def _prefetch_handle(
     state: _FSDPState,
-    current_handles_key: _HandlesKey,
+    current_handles_key: FlatParamHandle,
```
```suggestion
    current_handle: FlatParamHandle,
```
and downstream

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -1109,34 +1094,30 @@ def _prefetch_handles(
     state: _FSDPState,
-    current_handles_key: _HandlesKey,
-) -> List[_HandlesKey]:
+    current_handles_key: FlatParamHandle,
```
```suggestion
    current_handle: FlatParamHandle,
```

awgu reviewed (commented):
Thanks so much for grinding through this tedious change!

I made an initial pass through this.

voznesenskym commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -412,14 +411,14 @@ def _init_core_state(
-        nn.Module, List[FlatParamHandle]
-    ] = collections.defaultdict(list)
-    state._fully_sharded_module_to_handles = _fully_sharded_module_to_handles
+    _fully_sharded_module_to_handle: Dict[
+        nn.Module, FlatParamHandle
+    ] = collections.defaultdict(None)
```
Just change it to dict()?

awgu replied:
Yep!

voznesenskym commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -789,19 +789,25 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
+                kwargs,
             )
-            for handle in self._handles:
+            handle = self._handle
```
well, nothing really matters, you know?

voznesenskym commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -920,16 +926,19 @@ def _apply(self, *args, **kwargs):
         self,
-        *args,
-        **kwargs,
+        prefix: str = "",
```
Missed from another PR

voznesenskym committed:
Update on "Migrate tuple(handle) -> handle"




cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

wconstab commented:
would be good to write in the PR summary what the motivation/summary of the change is.

"We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping"

Related ux question though for @awgu - if we wanted to support a cleaner ux for, say, huge DHEN layers that need to be broken into multiple flatparams for perf reasons, would our best way to do that be by having a 1:many wrapping, or can we focus instead on better ways to map the model into FSDP-managed-modules?

voznesenskym committed:
Update on "Migrate tuple(handle) -> handle"


We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "Migrate tuple(handle) -> handle"


We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -1379,22 +1345,23 @@ def _register_post_backward_reshard_only_hooks(
+    hook_handle = register_multi_grad_hook(
+        inp_tensors, functools.partial(_post_backward_reshard, state, handle)
+    )
+    handle.flat_param._post_backward_hook_state = hook_handle  # type: ignore[attr-defined]
```
This should unblock the last unit test!

I tested via:
```
python -m pytest test/distributed/fsdp/test_fsdp_freezing_weights.py
```

voznesenskym committed:
Update on "Migrate tuple(handle) -> handle"


We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_unshard_params.py:
```diff
@@ -387,13 +387,16 @@ def _test_unshard_params_recurse(
         expected_outer_numel = (
             unsharded_outer_numel if unshard_outer else sharded_outer_numel
         )
+        for fsdp_module in FSDP.fsdp_modules(fsdp_model):
```
Did you insert this in the wrong place? (Should it be back to L607?)

(Test is failing since `fsdp_model` is not defined.)

voznesenskym replied:
Yea I think so lol

voznesenskym committed:
Update on "Migrate tuple(handle) -> handle"


We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_unshard_params.py:
```diff
@@ -288,8 +288,8 @@ def _test_unshard_params_respects_reshard(
-        outer_flat_param = model._handles[0].flat_param
-        inner_flat_param = model.module[0]._handles[0].flat_param
+        outer_flat_param = model._handle.flat_param
+        inner_flat_param = model.module._handle.flat_param
```
```suggestion
        inner_flat_param = model.module[0]._handle.flat_param
```

awgu commented:
> Related ux question though for @awgu - if we wanted to support a cleaner ux for, say, huge DHEN layers that need to be broken into multiple flatparams for perf reasons, would our best way to do that be by having a 1:many wrapping, or can we focus instead on better ways to map the model into FSDP-managed-modules?

Unfortunately, the 1:many design has two concerns: (1) how to construct the "many" is challenging to decide with only static information and (2) it introduces nontrivial complexity to our design. The latter part can be tackled, but the former part has been a major blocker. We explored this with our intern last summer, but we do not have a good solution at the moment for (1).

voznesenskym committed:
Update on "Migrate tuple(handle) -> handle"


We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_unshard_params.py:
```diff
@@ -289,6 +289,8 @@ def _test_unshard_params_respects_reshard(
             **fsdp_kwargs,
         )
         outer_flat_param = model._handle.flat_param
+        if not hasattr(model.module, "_handle"):
+            return
```
I did not quite follow why you added this.

voznesenskym committed:
Update on "Migrate tuple(handle) -> handle"


We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

wconstab commented:
> Unfortunately, the 1:many design has two concerns: (1) how to construct the "many" is challenging to decide with only static information and (2) it introduces nontrivial complexity to our design. The latter part can be tackled, but the former part has been a major blocker. We explored this with our intern last summer, but we do not have a good solution at the moment for (1).

Well, what if we give ourselves a break and don't try to solve (1) eagerly during eager?  Instead, we can expose an api that lets users specify the mapping between parameter fqns and fsdp units, and we could have an 'offline' utility that generates an optimal default by doing some sort of compiler pass or profiling, and lets users also edit/tweak before freezing the final dict to be used by fsdp?  Then in the actual 'compiled' flow, we could do this by default

voznesenskym committed:
Update on "Migrate tuple(handle) -> handle"


We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -1041,25 +1027,17 @@ def _catch_all_reshard(
-                # still points to the sharded data, so we need to reshard to
-                # use sharded views
-                and not handle._skipped_use_sharded_views
+                state._handle.flat_param.data_ptr()
+                == state._handle.flat_param._local_shard.data_ptr()
```
This is the fix for test_fsdp_sharded_grad_scaler.py:
```suggestion
                state._handle.flat_param.data_ptr()
                == state._handle.flat_param._local_shard.data_ptr()
                # If FSDP skipped using sharded views, then the flat parameter
                # still points to the sharded data, so we need to reshard to
                # use sharded views
                and not state._handle._skipped_use_sharded_views
```
(again looks like a versioning issue where some code got lost)

awgu replied:
Shall we preserve the comment?

awgu commented on test/distributed/fsdp/test_fsdp_unshard_params.py:
```diff
@@ -236,7 +236,7 @@ def test_unshard_singleton_param_writeback(self):
         """
         model = FSDP(nn.Linear(1, 1, bias=False, device=self.device))
-        flat_param = model._handles[0].flat_param
+        flat_param = model._handle[0].flat_param
```
```suggestion
        flat_param = model._handle.flat_param
```

awgu reviewed (commented):
I left some nits plus two fixes for failing unit tests.

voznesenskym committed:
Update on "Migrate tuple(handle) -> handle"


We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot rebase

voznesenskym committed:
Update on "Migrate tuple(handle) -> handle"


We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -441,57 +428,55 @@ def _pre_forward(
-
-        should_cast_forward_inputs = len(state._handles) > 0 and all(
-            not handle._force_full_precision for handle in state._handles
+        if handle._offload_params and handle.flat_param._cpu_grad is None:
```
```suggestion
        if handle and handle._offload_params and handle.flat_param._cpu_grad is None:
```

voznesenskym committed:
Update on "Migrate tuple(handle) -> handle"


We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "Migrate tuple(handle) -> handle"


We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

voznesenskym commented:
Only failing CI! @awgu can you do another quick pass? I'll address existing comments and linter for the next iteration.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -441,57 +428,55 @@ def _pre_forward(
 def _post_forward(
     state: _FSDPState,
-    handles: List[FlatParamHandle],
+    handle: FlatParamHandle,
```
```suggestion
    handle: Optional[FlatParamHandle],
```
(this one actually has some semantic importance since we should still run the post-forward logic, namely registering pre-backward hooks, even if the handle is `None` -- this matters for the root FSDP specifically right now)

awgu reviewed (approved):
LGTM! I left one more round of minor nits/fixes. Thanks so much for persisting on this.

Would it be possible to wait for https://github.com/pytorch/pytorch/pull/105346 to land since that unblocks some internal users (and external ones) and rebase on top of it? I do not expect any real rebase conflicts.

voznesenskym committed:
Update on "Migrate tuple(handle) -> handle"


We strengthen the invariant that one FSDP managed module has one flatparameter, and remove unused code that would have supported 1:many module to flatparam mapping


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot merge
